docs/design/018-new-http-api.md

```diff
@@ -0,0 +1,486 @@
+
+For example:
+```javascript
+import { Client, Request } from 'k6/x/net/http';
+
+export default async function () {
+  const client = new Client({
+    headers: { 'User-Agent': 'k6' },  // set some global headers
+  });
+  const request = new Request('https://httpbin.test.k6.io/get', {
+    // These will be merged with the Client options.
+    headers: { 'Case-Sensitive-Header': 'somevalue' },
+  });
+  const response = await client.get(request, {
+    // These will override any options for this specific submission.
+    headers: { 'Case-Sensitive-Header': 'anothervalue' },
+  });
+  const jsonData = await response.json();
+  console.log(jsonData);
+}
+```
+
+
+#### Data streaming
+
+The [Streams API](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API) allows streaming data that is received or sent over the network, or read from or written to the local filesystem. This enables more efficient usage of memory, as only chunks of it need to be allocated at once.
+
+This is a separate project from the HTTP API, tracked in [issue #2978](https://github.com/grafana/k6/issues/2978), and involves changes in other parts of k6. Certain HTTP API functionality, however, depends on this API being available.
+
+An example inspired by [Deno](https://deno.land/manual/examples/fetch_data#files-and-streams) of how this might work in k6:
+```javascript
+import { open } from 'k6/x/file';
+import { Client } from 'k6/x/net/http';
+
+// Will need supporting await in init context
+const file = await open('./logo.svg');  // by default assumes 'read'
+
+export default async function () {
+  const client = new Client();
+  await client.post('https://httpbin.test.k6.io/post', { body: file.readable });
+}
+```
+
+
+#### Fetch API
+
+The [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) is a convenience wrapper over existing Client, Socket and other low-level interfaces, with the benefit of being easy to use, and having sane defaults. It's a quick way to fire off some HTTP requests and get some responses, without worrying about advanced configuration.
+
+The implementation in k6 differs slightly from the web API, but we've tried to make it familiar to use wherever possible.
+
+Example:
+```javascript
+import { fetch } from 'k6/x/net/http';
+
+export default async function () {
+  await fetch('https://httpbin.test.k6.io/get');
+  await fetch('https://httpbin.test.k6.io/post', {
+    // Supports the same options as Client.request()
+    method: 'POST',
+    headers: { 'User-Agent': 'k6' },
+    json: { name: 'k6' },
+  });
+}
+```
+
+#### Events
+
+The new HTTP API will emit events which scripts can subscribe to, in order to implement advanced functionality.
+
+For example, a `requestToBeSent` event is emitted when the request was processed by k6, and just before it is sent to the server. This allows changing the request body or headers, or introducing artificial delays.
```

Comment:
I think this might be a way to implement features like #2667 and #1716. Let me know what you think.

It could also serve as a starting point for a global event system, which could be useful for #2432.